target_chains/ton/contracts/contracts/common/merkle_tree.fc

```diff
@@ -0,0 +1,57 @@
+#include "../imports/stdlib.fc";
+#include "utils.fc";
+
+const int MERKLE_LEAF_PREFIX = 0;
+const int MERKLE_NODE_PREFIX = 1;
+
+int to_u160(int value) {
+    ;; (1 << 160) creates a number with only the 161st bit set to 1 and all other bits set to 0.
+    ;; Subtracting 1 from this number flips all the bits below the 161st bit to 1, creating a mask of 160 1's.
+    ;; Finally, the & operation with this mask keeps only the 160 least significant bits of the input value and sets all other bits to 0.
+    int mask = (1 << 160) - 1;
+    return value & mask;
+}
+
+int leaf_hash(slice message) {
+    int hash = keccak256_slice(begin_cell()
+        .store_uint(MERKLE_LEAF_PREFIX, 8)
+        .store_slice(message)
+    .end_cell().begin_parse());
+    return hash >> 96;
+}
+
+int node_hash(int a, int b) {
+    int min_value = min(a, b);
+    int max_value = max(a, b);
+    int hash = keccak256_slice(begin_cell()
+        .store_uint(MERKLE_NODE_PREFIX, 8)
+        .store_uint(to_u160(min_value), 160)
+        .store_uint(to_u160(max_value), 160)
```

Comment:
why would you need `to_u160`? aren't these u160 by default when you read them?

Reply:
yeah good point will remove